components/core/src/transformers/ScopeInjection.scala

```diff
@@ -31,48 +31,39 @@ trait ScopeInjection extends MacroModule with TransformationUtils {
       newTree
     }
 
+    // Note: the Scala compiler sometimes leaves ".this" in the path and we remove it.
+    def injectModule(t: Tree): Tree =
+      Ident(TermName(t.toString.replaceAll("\\.this", "").replaceAll("`package`", "package")).encodedName)
+
+    /*
+     * Translation rules:
+     *   [[x.y.z.obj.method]] ~> this.`x.y.z.obj`.method
+     */
     override def transform(tree: Tree): Tree = {
       log(" " * ident + " --> " + tree, 3)
       ident += 1
 
       val result = tree match {
-        //provide Def trees with NoSymbol (for correct show(tree))
-        case vdDef: ValOrDefDef if rewireThis => {
-          val retDef = super.transform(tree)
-          retDef.setSymbol(NoSymbol)
-          retDef
-        }
-
-        case CaseDef(pat: Tree, guard: Tree, body: Tree) => {
-          val newPat = pat match {
-            case Apply(fun: TypeTree, args) => transform(preservePosition(Apply(fun.original, args), pat))
-            case _                          => transform(pat)
-          }
-          CaseDef(newPat, transform(guard), transform(body))
-        }
-
-        // re-wire objects
-        case s @ Select(Select(inn, t: TermName), name) // package object goes to this
-        if s.symbol.isMethod && (rewiredToThis(t.toString) || t.toString == "this") =>
-          Ident(name)
-
-        case s @ Select(inn, name) if s.symbol.isMethod =>
-          Select(transform(inn), name)
+        case Apply(Select(This(typeNames.EMPTY), TermName("lift")), _) =>
+          tree
 
-        // replaces objects with their cake counterparts
-        case s @ Select(inn, name) if s.symbol.isModule =>
-          Ident(name)
+        case s @ Select(inn, name) if inn.symbol.isPackage && s.symbol.isModule =>
+          injectModule(s)
+
+        case Select(inn, name) if inn.symbol.isModule =>
+          Select(transform(inn), name)
 
-        // Added to rewire inherited methods to this class
-        case th @ This(_) if rewireThis =>
-          This(typeNames.EMPTY)
+        //Optimization: adds implicit arguments manually.
+        case MultipleTypeApply(lhs @ Select(inn, name), targs, argss
+          ) if !inn.symbol.isModule && !isFunction(lhs.symbol) =>
+          val newqqc = transform(inn)
+          val tlhs = if (inn.tpe != null)
```

Comment:
It would be great to have a flag for specifying if we would like to make implicits explicit or not.